### PR TITLE
Alarm Syscall: fix computed `dt` when both unshifted `reference` and `dt` from userspace have low-order bits.

### PR DESCRIPTION
**Pull request by alevy**
### Pull Request Overview

This pull request fixes two issues in the alarm syscall driver, both necessary to correctly handle alarms that overflow in userspace. Each issue separated into its own commit:

1. The second argument to the alarm upcall is the time the alarm was scheduled to expire. This PR fixes it to be left-justified and thus expressed in the same units as all other counter values exposed to userspace.
2. Computing `dt` for a userspace alarm when userspace values are left-shifted to be u32 (e.g. on the 24-bit nRF52 RTC implementation) underestimated `dt` in some cases where both `reference` and `dt` had low-order bits that were rounded off. It is necessary to compute `dt` _considering_ the `reference` value when it is provided, so this PR computes `dt` from the computed expiration in the shifted domain.

### Testing Strategy

The second commit adds test cases for 24-bit alarms with provided reference counters. Running `cargo test` without the changes in the second commit fail, and succeed with the changes. The first commit is tested via changes to libtock-c in a different PR to that repo.

### TODO or Help Wanted

It would be good to be able to test upcall values in unit tests, but I decided to avoid figuring out how to do that for this PR since the relevant change is pretty trivially correct, and was unnoticed simply because that value was never used.

### Documentation Updated

- [X] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [X] Ran `make prepush`.

### CODE REVIEW

File: capsules/core/src/alarm.rs [hunk 1]
```diff
@@ -877,6 +893,58 @@ mod test {
...
+        assert!(Ticks24::u32_padding() == 8);
+
+        let armed_time =
+            AlarmDriver::<MockAlarm<Ticks24, Freq10MHz>>::rearm_u32_left_justified_expiration(
+                // Current time:
+                Ticks24::from(6_u32),
+                // Userspace-provided reference:
+                Some(Ticks24::from(5_u32).into_u32_left_justified() - 43),
+                // Left-justified `dt` value:
+                Ticks24::from(2_u32).into_u32_left_justified() + 43,
```
Comment (lschuermann) [inline capsules/core/src/alarm.rs:935]:
```suggestion
                // Userspace-provided reference, including bits not representable in Ticks24:
                // (5 << 8) - 43 = 1237
                Some(Ticks24::from(5_u32).into_u32_left_justified() - 43),
                // Left-justified `dt` value, including bits not representable in Ticks24:
                // (2 << 8) - 43 = 469
                Ticks24::from(2_u32).into_u32_left_justified() + 43,
```

File: capsules/core/src/alarm.rs [hunk 2]
```diff
@@ -877,6 +893,58 @@ mod test {
...
+                // Reference to the `Option<Expiration>`, also used
+                // to update the counter of armed alarms:
+                &mut expiration,
+            );
+
+        let expiration = expiration.unwrap();
+
+        assert_eq!(armed_time, 7 << 8);
```
Comment (lschuermann) [inline capsules/core/src/alarm.rs:943]:
```suggestion
        // When we naively round down reference `(1237 / 256 ~= 4.83 -> 4)` and
        // round up dt `(469 / 256 ~= 1.83 -> 2)` we'd arm the alarm to
        // `2 + 4 = 6`. However, when considering the full resolution
        // `reference + dt` `(1237 + 256) / 256 ~= 6.67` we can see that arming
        // to `6` will have the alarm fire too early. The alarm rearm code needs
        // to compensate for the case that (reference + dt) overflow and generate
        // a dt from that rounded value, in this case `7`.
        assert_eq!(armed_time, 7 << 8);
```

**Review by lschuermann — DISMISSED** [review]

File: capsules/core/src/alarm.rs [hunk 3]
```diff
@@ -235,10 +240,20 @@ impl<'a, A: Alarm<'a>> AlarmDriver<'a, A> {
...
+                    // firing _before_ the time requested by userspace. Add one to
+                    // compensate this:
+                    (expiration_shifted >> A::Ticks::u32_padding()) + 1
+                } else {
+                    expiration_shifted >> A::Ticks::u32_padding()
+                };
 
-        // Round dt up:
-        let dt_unshifted = if dt_u32 & ((1 << A::Ticks::u32_padding()) - 1) != 0 {
+            expiration_unshifted.wrapping_sub(reference_u32 >> A::Ticks::u32_padding())
```
[the remark this answers is not in the document]
Comment (alevy) [reply capsules/core/src/alarm.rs]: Added a comment

**Review by lschuermann — APPROVED** [review]

**Comment by alevy** [comment]
@lschuermann last one I promise, just fixes formatting

**Review by bradjc — APPROVED** [review]

**Comment by alevy** [comment]
Agree, I don't there are anything compatibility issues

**Review by tyler-potyondy — APPROVED** [review]

**Pull request merged**